### ISSUE TITLE
Migration doesn't remove undeclared schemas

### ISSUE THREAD

**Issue by victoriafrench** [realm/realm-js#573]
Not sure if this is a bug (though I feel it should be) and it's driving me crazy.

So I have my v1 schema that has let's say ['Dog', 'Cat']
Then I have v2 schema (with migration) lets say ['Cat', 'Mouse]

Migration works great... but the database now contains Dog, Cat, and Mouse. I would think it would now only be Cat and Mouse as I am no longer declaring Dog as part of the schema. But the table is still there.

I see in the iOS, Java, and Swift kit there is the capability to delete a table during migration. But none of those examples work for react-native. So how is this done or when will it be available.

Yes I did scour the internet for the past two days trying to find an answer and 'delete your database' is not acceptable as an answer.

Thanks.

**Comment by victoriafrench** [comment]
So here is a real test scenario with interesting results:

```javascript
import Realm from 'realm';

const KeyCode_v1 = {
  name: 'KeyCode',
  primaryKey: 'code',
  properties: {
    code: 'string',
    formid: { type: 'string', indexed: true }
  }
};

const FormInfo_v1 = {
  name: 'FormInfo',
  primaryKey: 'id',
  properties: {
    id: 'string',
    deleted: 'bool',
    locked: 'bool',
    name: 'string',
    keycode: 'string'
  }
};

const KeyCodes = {
  name: 'KeyCodes',
  primaryKey: 'keycode',
  properties: {
    keycode: 'string',
    name: 'string',
    formsec: 'string',
    locked: 'bool',
    deleted: 'bool'
  }
};

function migrateV2(oldRealm, newRealm) {
  if (oldRealm.schemaVersion < 2) {
    var oldFormInfo = oldRealm.objects('FormInfo');
    var newKeyCodes = newRealm.objects('KeyCodes');
    for (let i = 0; i < oldFormInfo.length; i++) {
      let wf = oldFormInfo[i];
      let nf = {
        keycode: wf.keycode,
        name: wf.name,
        formsec: wf.id,
        locked: wf.locked || false,
        deleted: wf.deleted || false
      };
      let index = newRealm.objects('KeyCodes').filtered(`keycode == "${nf.keycode}"`);
      if ( index.length === 0 ) {
        newRealm.create('KeyCodes', nf);
      } else if (nf.locked === true && index[0].locked === false) {
        // record the new item only if it is locked and the other one isn't
        newRealm.delete(index);
        newRealm.create('KeyCodes', nf);
      }
    }
  }
};

function migrateV3(oldRealm, newRealm) {
  if (oldRealm.schemaVersion < 3) {
    newRealm.deleteAll();
    let keycodes = oldRealm.objects('KeyCodes');
    for (let i = 0; i < keycodes.length; i++) {
      newRealm.create('KeyCodes', keycodes[i]);
    }
  }
}

function migrateV4(oldRealm, newRealm) {
  if (oldRealm.schemaVersion < 4) {
    newRealm.deleteAll();
    let keycodes = oldRealm.objects('KeyCodes');
    for (let i = 0; i < keycodes.length; i++) {
      newRealm.create('KeyCodes', keycodes[i]);
    }
  }
}

function migrateV6(oldRealm, newRealm) {
  if (oldRealm.schemaVersion < 6) {
    let keycodetable = newRealm.objects('KeyCode');
    for (let i = 0; i < keycodetable.length; i++) {
      newRealm.delete(keycodetable[i]);
    }
    let formtable = newRealm.objects('FormInfo');
    for (let i = 0; i < formtable.length; i++) {
      newRealm.delete(formtable[i]);
    }
  }
}

function migrateV8(oldRealm, newRealm) {
  if (oldRealm.schemaVersion < 8) {
    let keycodetable = newRealm.objects('KeyCode');
    newRealm.delete(keycodetable);
    let formtable = newRealm.objects('FormInfo');
    newRealm.delete(formtable);
  }
}

const schemas = [
  { schema: [KeyCode_v1, FormInfo_v1], schemaVersion: 1},
  { schema: [KeyCodes], schemaVersion: 2, migration: migrateV2} ,
  { schema: [KeyCodes], schemaVersion: 3, migration: migrateV3 },
  { schema: [KeyCode_v1, FormInfo_v1, KeyCodes], schemaVersion: 4, migration: migrateV4 },
  { schema: [KeyCodes], schemaVersion: 5 },
  { schema: [KeyCode_v1, FormInfo_v1, KeyCodes], schemaVersion: 6, migration: migrateV6 },
  { schema: [KeyCodes], schemaVersion: 7 },
  { schema: [KeyCode_v1, FormInfo_v1, KeyCodes], schemaVersion: 8, migration: migrateV8 },
  { schema: [KeyCodes], schemaVersion: 9 }
];

/* Preform migrations */
let nextSchemaIndex = Realm.schemaVersion(Realm.defaultPath);
while(nextSchemaIndex < schemas.length) {
  let migratedRealm = new Realm(schemas[nextSchemaIndex++]);
  migratedRealm.close();
}

const realm = new Realm(schemas[schemas.length - 1]);
export default realm;
```

- `migrageV2` left `KeyCode` and `FormInfo` and all the data in the migrated database. Expected that the tables would be dropped.
- `migrateV3` had no effect at all. Expected the whole database to be empty and a new `KeyCodes` table added. It resulted in the same as V2.
- `migrateV4` was an attempt at the V3 but with the tables in the schema. Same results, nothing happened.
- `migrateV6` it did start deleting data, but left 4 items in `KeyCode` and 3 items in `FormInfo` and I have no idea why.
- `migrateV8` finally removed all data from `KeyCode` and `FormInfo` but again the tables are left behind.

What was expected it that there was a way to completely remove `KeyCode` and `FormInfo` and migrate the data to `KeyCodes`.

**Comment by alazier** [comment]
We will need to add this feature to the react-native api in order to support this.

**Comment by ashwinphatak** [comment]
Merged 532f517ded0673c142d1b61ffe1b5de0822543fd, which exposes a `deleteModel` method to be used during migrations.
